specs: vep-2448 detailed design section (#2558)
vmware/versatile-data-kit: specs/vep-2448-vdk-run-logs-simplified-and-readable/README.md

The user configuration should take precedence over the global one.

User should be able to do log.info("xxx") inside a data job and vdk
automatically add structure around it.

Plugins should be able to declare and inject new metadata fields, e.g. DAGs
should be able to add a dag_name field or a current_job field

The above requirements are more than likely achievable using structlog.

https://www.structlog.org/en/stable/

### Log Less

We should make sure that logs throughout vdk have the correct log level. Any
logs that don't should be moved to the appropriate level, e.g. we seem to have
lots of info logs that should be on the debug level.

Users have the ability to to change the log level for specific modules in their
data job with the `log_level_module` option. We should add documentation around
this.

### Clean Error Handling

Error messages should clearly state the problem without placeholder and repeated
logs text so that users can directly understand what went wrong. Users should be
able to see the original exception when it's passed up the call stack and is in
the user code so they can handle it.

> VDK developers should be discouraged to use generic error messages like "An

Reviewer: ```suggestion
VDK developers should be discouraged from using generic error messages like "An
```
nit

> error occurred". This will give more meaningful feedback to users.

Reviewer 2: Another idea potenitally. 

For some common errors we can consider creating an error dictionary/repository that stores all the known errors, their descriptions, and potential solutions.

Author: Added to https://github.com/vmware/versatile-data-kit/issues/2571

### Progress Indicators

> Data jobs in the CLI should display progress indicators by default instead of

Reviewer 2: We should consider implementing an asynchronous mechanism for updating progress indicators. This would make sure that the primary task doesn’t get interrupted or slowed down due to progress bar updates. This is really relevant for ingestion. 

Also as we need progress indicators in multiple places likely we need a common encapsulation for that - in a form of a python module or plugin. 

The progress indicator should providing a notification callback mechanism or similar so it's integratable with Notebook

Author: Added this as separate ticket https://github.com/vmware/versatile-data-kit/issues/2593

## Implementation stories

### Log Structure

1. [Spike] Research structlog and integration with VDK
   - How will log = logging.get_logger() work from now on
   - How would user logging work,
   - how would fields be configured and added
1. Add configuration options for logging metadata to config.ini
2. Display possible metadata fields to be logged as part of vdk help
3. Add possible metadata fields to be logged in the documentation

### Log Less

1. Go through all the logs for vdk-core and plugins to check if they're at the
   appropriate level
2. Test and document `log_level_module`

### Clean Error Handling

1. Get rid of the log and throw/rethrow pattern
   - recoverable exceptions should be logged at the warn level and swallowed
   - unrecoverable exceptions should be either wrapped in a generic error and
     re-thrown or just thrown and passed to the caller
2. Create specific error classes for User, Platform and Config errors
   - possibly extend the classes into more specific errors, e.g. DataQualityException
   - the logic that determines the type of error and who's responsible for
     fixing it, e.g. get_blamee should remain in place

>    - developers should have the option to pass the error up to the data job

Reviewer 2: There's need for more indepth proposal here. Let's consider a spike for that and think it through.

The catch is we want to propagate an original exception up the call stack . But we also want to keep information about its classification (e.g., "user error", platform error). 

So how do we do that ? 

Options 1 Define Custom Exception class with attributes user to use:

We raise an exception like that 

```
class VDKUserCodeException(Exception):
    def __init__(self, message, original_exception=None):
        super().__init__(message)
        self.original_exception = original_exception
```

Or we can use the native python `__cause__` and exception chaining https://peps.python.org/pep-3134/ 
But we need to make it somewhat mandatory

Specifically for user vs platform error I was thinking having it be attribute might be more appropriate since the same FileNotFound exception for example could be user or platform depending on where it was raised from:  
```
class VDKException(Exception):
    def __init__(self, message, resolvable_class : ResolveableBy=None, original_exception=None):
        super().__init__(message)
        self.resolvable_class = resolvable_class 
        self.original_exception = original_exception
```

And user would use it like that 
```
try: 
   job_input.execue_template() 
except VDKException e: 
   if isinstance(e.original_exception, DataQualityException): 
      do_someting() 
 ```
  This is not very intuitive UX in ptyhon though. 
  
  
  I think users would except something like 
```
try: 
   execute_risky_operation()
except OriginalException as e:
  ... 
```

If we want that while still providing a mechanism to classify the exception we may add custom attributes: 
```
def classify_exception(exception, classification):
    exception.classification = classification
    return exception
```

Author: Linked to https://github.com/vmware/versatile-data-kit/issues/2572
